gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java

```diff
@@ -309,11 +315,86 @@ public ChangeRequest getChangeRequest(final String zoneName, final String change
+    }
+  }
+
+  /**
+   * Since {@code BatchRequest} is a final class, it cannot be mocked with easy mock and the call of
+   * {@code execute()} cannot be tested. Thus, most of the functionality of {@link
+   * #submitBatch(DnsBatch)} is extracted to this method which does not make the call so it does not
+   * communicate with the service.
+   */
+  BatchRequest prepareBatch(DnsBatch toSubmit) throws IOException {
+    BatchRequest batch = null;
+    for (Map.Entry<DnsBatch.Request, DnsBatch.Callback> entry : toSubmit.requests().entrySet()) {
+      DnsBatch.Request request = entry.getKey();
+      DnsBatch.Callback callback = entry.getValue();
+      switch (request.operation()) {
+        case LIST_ZONES:
+          JsonBatchCallback rpcCallback = listZonesCallback(callback, request);
+          batch =
+              dnsRpc.enqueueListZones(batch, request, rpcCallback, optionMap(request.options()));
+          break;
+        default:
+          // todo(mderka) implement the rest of the operations
+          throw new UnsupportedOperationException("Not implemented yet");
+      }
+    }
+    return batch;
+  }
+
+  @Override
+  public DnsBatch batch() {
+    return new DnsBatch(this);
+  }
+
+  private JsonBatchCallback listZonesCallback(final DnsBatch.Callback callback,
+      final DnsBatch.Request request) {
+    return new JsonBatchCallback<ManagedZonesListResponse>() {
+      @Override
+      public void onFailure(GoogleJsonError error, HttpHeaders httpHeaders) {
+        if (callback != null) {
+          callback.error(new DnsException(error), request);
+        }
+      }
+
+      @Override
+      public void onSuccess(ManagedZonesListResponse zoneList, HttpHeaders httpHeaders) {
+        if (callback != null) {
+          DnsRpc.ListResult<ManagedZone> listResult =
+              DnsRpc.ListResult.of(zoneList.getNextPageToken(), zoneList.getManagedZones());
+          String cursor = listResult.pageToken();
+          Iterable<Zone> zones = listResult.results() == null ? ImmutableList.<Zone>of()
+              : Iterables.transform(listResult.results(), new Function<ManagedZone, Zone>() {
+                @Override
+                public Zone apply(ManagedZone managedZone) {
+                  return new Zone(options().service(),
+                      new ZoneInfo.BuilderImpl(ZoneInfo.fromPb(managedZone)));
+                }
+              }
+          );
+          Page<Zone> page = new PageImpl<>(
+              new ZonePageFetcher(options(), cursor, optionMap(request.options())), cursor, zones);
+          callback.success(page, request);
+        }
+      }
+    };
+  }
+
   private Map<DnsRpc.Option, ?> optionMap(AbstractOption... options) {
     Map<DnsRpc.Option, Object> temp = Maps.newEnumMap(DnsRpc.Option.class);
-    for (AbstractOption option : options) {
-      Object prev = temp.put(option.rpcOption(), option.value());
-      checkArgument(prev == null, "Duplicate option %s", option);
+    if (options != null) {
```

Comment:
Can this ever happen? If you call it using `optionMap(request.options())` the getter already takes care of not returning `null`.